Add Python 3 support

mx-moth opened the pull request:
Each commit is self contained and hopefully gives enough context to be understandable. The end result is Python 3.5 support.

mx-moth committed:
Add Python 3 to tox.ini

Django is not compatible with all versions of Python 3 across all
versions. tox will only test compatible versions.

mx-moth committed:
Use BeautifulSoup4 for py3 compatibility

BeautifulSoup4 deprecated the `BeautifulStoneSoup` class, preferring
`BeautifulSoup(xml_string, features="xml")` instead. `lxml` is required
to use BeautifulSoup4 in xml-mode, so this was installed. `lxml` (and
XML in general) is case-sensitive, so some areas where attributes were
accessed in lowercase were fixed. `lxml` is also more strict than the
old XML parsing used in BS3, dropping namespace prefixes for unknown
namespaces. The `saml` namespace was missing from some assertions,
causing things to fail. The `xmlns:saml` declaration has been moved to
the root node of the document to fix this.

mx-moth committed:
Use print as a function

mx-moth committed:
Use py3-style relative imports

mx-moth committed:
Update import paths / function names for py3

mx-moth committed:
bytes vs. strings

Python 2 can mix and match bytes and strings quite a lot and everything
works - until it doesnt. Python 3 is much stricter, and made clear some
points where bytes were being used as strings, and vice-versa.

Where appropriate, strings have been encoded using 'utf8', as this is
the encoding used in the generated XML documents.

mx-moth committed:
Use pytest-cov over pytest-cover

pytest-cover was merged back in to pytest-cov, and is no longer
maintained. pytest-cov is compatible with Python 3, pytest-cover is not.

mx-moth committed:
Fix tests to reflect case sensitive, restructured XML

mx-moth committed:
Remove pdbpp from development requirements

It does not install well on Python 3, and is not required to run the
tests or develop. Developers who would like to use pdbpp in development
are welcome to install it manually as the require it.

mx-moth committed:
Add Python 3 environments to travis config

danni commented on idptest/tests/test_google_apps.py:
```diff
@@ -7,12 +7,13 @@
     'ProviderName="google.com" IsPassive="false" '
     'AssertionConsumerServiceURL="https://www.google.com/a/example.com/acs">'
-    '<saml:Issuer xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">'
+    '<saml:Issuer>'
```
This request is generated by the SP right? In which case, can we change it?

I don't understand why we can't have the namespace on the `saml` object anyway, since it's not used anywhere else.

mx-moth replied:
Good point. I may have been a touch enthusiastic with my find/replace here.

danni commented on saml2idp/xml_templates.py:
```diff
@@ -71,7 +71,7 @@
 # Minimal assertion for Google Apps:
 ASSERTION_GOOGLE_APPS = (
-    '<saml:Assertion xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" '
+    '<saml:Assertion '
```
Ditto above.

mx-moth replied:
This assertion forms part of the response, so we can change this one.

danni commented on saml2idp/xml_templates.py:
```diff
@@ -92,7 +92,7 @@
 
 # Minimal assertion for SalesForce:
 ASSERTION_SALESFORCE = (
-    '<saml:Assertion xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" '
```
Ditto above.

mx-moth replied:
This also forms part of the response, so it should be fine to change.

danni commented on saml2idp/xml_templates.py:
```diff
@@ -118,12 +118,13 @@
                     'IssueInstant="${ISSUE_INSTANT}" '
                     'Version="2.0">'
-        '<saml:Issuer xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">${ISSUER}</saml:Issuer>'
+        '<saml:Issuer>${ISSUER}</saml:Issuer>'
```
However, this is our response right, in which case this one is fine :)

danni commented on idptest/tests/test_salesforce.py:
```diff
@@ -4,7 +4,7 @@
 # standard library imports:
```
`from future import absolute_imports`

mx-moth replied:
I'll make a new commit that adds `absolute_imports, print_function, unicode_literals` to all the files. That covers all the **future** imports for Python 2 except division.

danni replied:
May as well do division too. That prevents any weird surprises.

danni commented on saml2idp/demo.py:
```diff
@@ -1,6 +1,6 @@
-import base
```
Ditto.

danni commented on saml2idp/google_apps.py:
```diff
@@ -1,7 +1,7 @@
-import base
```
Ditto.

danni commented on saml2idp/metadata.py:
```diff
@@ -4,7 +4,7 @@
 # Django imports
```
Ditto.

danni commented on saml2idp/salesforce.py:
```diff
@@ -1,6 +1,6 @@
-import base
```
Ditto.

danni commented on saml2idp/shib.py:
```diff
@@ -1,5 +1,5 @@
-import base
```
Ditto.

danni commented on saml2idp/urls.py:
```diff
@@ -1,6 +1,6 @@
 from django.conf.urls import patterns, url
```
Ditto.

danni commented on setup.py:
```diff
@@ -22,7 +22,7 @@
         'Django>=1.4',
         # We have to pin M2Crypto to version 0.22.3 because more recent
         # versions are failing due to issues with finding openssl libs.
```
This comment is wrong now.

danni commented on idptest/tests/test_salesforce.py:
```diff
@@ -14,8 +14,9 @@
-    '<saml:Issuer xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">'
+    'Version="2.0" xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" '
+    'xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">'
+    '<saml:Issuer>'
```
This is a request from the app right? In which case we can't change it.

mx-moth replied:
Yup, I should not change this one.

danni commented on idptest/tests/test_signing.py:
```diff
@@ -33,8 +33,8 @@
 }
 
-ASSERTION_SALESFORCE_XML = '<saml:Assertion xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_7ccdda8bc6b328570c03b218d7521772998da45374" IssueInstant="2011-08-11T23:38:34Z" Version="2.0"><saml:Issuer>http://127.0.0.1:8000</saml:Issuer><saml:Subject><saml:NameID Format="urn:oasis:names:tc:SAML:2.0:nameid-format:email" SPNameQualifier="example.net">randomuser@example.com</saml:NameID><saml:S ...
-SIGNED_ASSERTION_SALESFORCE_XML = '<saml:Assertion xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_7ccdda8bc6b328570c03b218d7521772998da45374" IssueInstant="2011-08-11T23:38:34Z" Version="2.0"><saml:Issuer>http://127.0.0.1:8000</saml:Issuer><ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#"><ds:SignedInfo><ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14 ...
```
Ditto.

mx-moth replied:
These ones form part of the response, so should be changeable.

danni commented on idptest/tests/test_signing.py:
```diff
@@ -47,10 +47,10 @@
+RESPONSE_XML = '<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" Destination="https://www.example.net/a/example.com/acs" ID="_2972e82c07bb5453956cc11fb19cad97ed26ff8bb4" InResponseTo="mpjibjdppiodcpciaefmdahiipjpcghdcfjodkbi" IssueInstant="2011-08-11T23:38:34Z" Version="2.0"><saml:Issuer>http://127.0.0.1:8000</saml:Issuer><samlp ...
+RESPONSE_WITH_SIGNED_ASSERTION_SALESFORCE_XML = '<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" Destination="https://www.example.net/a/example.com/acs" ID="_2972e82c07bb5453956cc11fb19cad97ed26ff8bb4" InResponseTo="mpjibjdppiodcpciaefmdahiipjpcghdcfjodkbi" IssueInstant="2011-08-11T23:38:34Z" Version="2.0"><saml:Issuer>http://1 ...
 
-SIGNED_RESPONSE_WITH_SIGNED_ASSERTION_SALESFORCE_XML = '<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" Destination="https://www.example.net/a/example.com/acs" ID="_2972e82c07bb5453956cc11fb19cad97ed26ff8bb4" InResponseTo="mpjibjdppiodcpciaefmdahiipjpcghdcfjodkbi" IssueInstant="2011-08-11T23:38:34Z" Version="2.0"><saml:Issuer xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">h ...
```
Ditto.

danni commented on .travis.yml:
```diff
@@ -1,20 +1,28 @@
-    packages:
-      - swig
+  - 2.7
+  - 3.5
```
FWIW we're running 3.4 in prod atm.

mx-moth replied:
I've added Python 3.4 to tox and travis.

mx-moth committed:
Add sensible __future__ imports to all files

This exposed some more lingering bytes vs. str issues, especially one
function in pyopenssl that requires bytes in py2, and str in py3.

danni commented:
This generally looks pretty good to me now.

timheap commented:
Cool, I'll send it to upstream for review then. Thanks!
